File: experiments/java-sdk/examples/com/datadog/llmobs/experiments/examples/TopicRelevance.java
package com.datadog.llmobs.experiments.examples;

import com.datadog.llmobs.experiments.Dataset;
import com.datadog.llmobs.experiments.Evaluator;
import com.datadog.llmobs.experiments.Experiment;
import com.datadog.llmobs.experiments.ExperimentResult;
import com.datadog.llmobs.experiments.ExperimentsClient;
import com.datadog.llmobs.experiments.Task;

import java.util.Map;

/**
 * End-to-end example: create a dataset, run a task per record, ship results to Datadog.
 *
 * <p>Reads {@code DD_API_KEY}, {@code DD_APPLICATION_KEY}, and (optional) {@code DD_SITE} from
 * the environment. The task here is a mock keyword-overlap classifier so the example needs no
 * external LLM dependency — swap in a real OpenAI / Anthropic call for your own use.
 *
 * <p>Run with: {@code ./gradlew run}
 */
public final class TopicRelevance {

    public static void main(String[] args) {
        String apiKey = requireEnv("DD_API_KEY");
        String appKey = requireEnv("DD_APPLICATION_KEY");
        String site = System.getenv().getOrDefault("DD_SITE", "datadoghq.com");

        ExperimentsClient client = ExperimentsClient.builder()
            .apiKey(apiKey)
            .applicationKey(appKey)
            .site(site)
            .projectName("java-sdk-bootstrap")
            .build();

        long ts = System.currentTimeMillis();
        // Each record can carry optional metadata as a Map<String, Object>. Metadata is stored
        // on the dataset record in Datadog and is visible in the dataset view of the dashboard.
        // The span for each row links back to the record via the dataset_record_id tag, so the
        // UI can surface this metadata when drilling into a row's results.
        Dataset dataset = client.createDataset("topic-relevance-demo-" + ts, "Java SDK v0.1 smoke test")
            .addRecord(
                Map.of("prompt", "I love hiking in the mountains on weekends.", "topics", "outdoor, travel"),
                "true",
                Map.of("source", "synthetic", "difficulty", "easy")
            )
            .addRecord(
                Map.of("prompt", "Explain quantum entanglement in two sentences.", "topics", "outdoor, travel"),
                "false",
                Map.of("source", "synthetic", "difficulty", "easy")
            )
            .addRecord(
                Map.of("prompt", "Best Italian restaurants in Brooklyn?", "topics", "food, nyc"),
                "true",
                Map.of("source", "user-report", "difficulty", "medium", "reviewer", "alex")
            )
            .addRecord(
                Map.of("prompt", "How do I configure nginx for HTTPS?", "topics", "food, nyc"),
                "false",
                Map.of("source", "user-report", "difficulty", "hard", "reviewer", "alex")
            );

        Task<Map<String, Object>, Map<String, Object>> classify = (input, config) -> {
            String prompt = (String) input.get("prompt");
            String topics = (String) input.get("topics");
            boolean overlap = keywordOverlap(prompt, topics);
            return Map.of(
                "response", String.valueOf(overlap),
                "confidence", overlap ? 0.85 : 0.65
            );
        };

        Evaluator<Boolean> exactMatch = (input, output, expected) -> {
            Map<?, ?> out = (Map<?, ?>) output;
            return out.get("response").equals(expected);
        };

        Evaluator<Double> confidenceScore = (input, output, expected) -> {
            Map<?, ?> out = (Map<?, ?>) output;
            return ((Number) out.get("confidence")).doubleValue();
        };

        Evaluator<String> verdict = (input, output, expected) -> {
            Map<?, ?> out = (Map<?, ?>) output;
            return "true".equals(out.get("response")) ? "in-topic" : "off-topic";
        };

        Experiment<Map<String, Object>, Map<String, Object>> exp =
            Experiment.<Map<String, Object>, Map<String, Object>>builder(client)
                .name("topic-relevance-demo")
                .dataset(dataset)
                .task(classify)
                .evaluator("exact_match", exactMatch)
                .evaluator("confidence_score", confidenceScore)
                .evaluator("verdict_category", verdict)
                .config(Map.of("approach", "keyword-overlap", "version", "v0.1"))
                .tags(Map.of("variant", "java-v0.1", "owner", "design-partner-bootstrap"))
                .build();

        ExperimentResult result = exp.run();
        System.out.println();
        System.out.println("Experiment URL : " + result.url());
        System.out.println("Experiment ID  : " + result.experimentId());
        System.out.println("Rows           : " + result.rows().size());
        for (ExperimentResult.Row row : result.rows()) {
            System.out.println("  row " + row.index()
                + " status=" + (row.isError() ? "error" : "ok")
                + " evals=" + row.evaluations());
        }
    }

    private static boolean keywordOverlap(String prompt, String topics) {
        String p = prompt.toLowerCase();
        for (String topic : topics.split(",")) {
            for (String word : topic.trim().toLowerCase().split("\\s+")) {
                if (!word.isEmpty() && p.contains(word)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static String requireEnv(String name) {
        String v = System.getenv(name);
        if (v == null || v.isBlank()) {
            System.err.println("Missing required env var: " + name);
            System.exit(1);
        }
        return v;
    }
}

File: experiments/node-sdk/examples/topicRelevance.ts
/**
 * TopicRelevance — direct port of the Java SDK's `TopicRelevance` example.
 *
 * Identical to the Java example in every respect — same four dataset records
 * (inputs, expected outputs, metadata), same evaluators, config and tags — except
 * it writes to the Node-specific project `node-sdk-bootstrap` (the Java example
 * uses `java-sdk-bootstrap`), so the two SDKs' artifacts stay in separate projects
 * for easy side-by-side comparison.
 *
 *   DD_API_KEY=... DD_APPLICATION_KEY=... [DD_SITE=datadoghq.com] npm run runTopicRelevance
 */
import { pathToFileURL } from "node:url";
import { Dataset, Experiment, ExperimentsClient } from "../src/index.js";

function requireEnv(name: string): string {
  const v = process.env[name];
  if (!v || v.trim() === "") {
    console.error(`Missing required env var: ${name}`);
    process.exit(1);
  }
  return v;
}

/**
 * Mock "task": does the prompt contain any keyword from the (comma-separated)
 * topics list? Mirrors the Java `keywordOverlap` exactly.
 */
function keywordOverlap(prompt: string, topics: string): boolean {
  const p = prompt.toLowerCase();
  for (const topic of topics.split(",")) {
    for (const word of topic.trim().toLowerCase().split(/\s+/)) {
      if (word !== "" && p.includes(word)) return true;
    }
  }
  return false;
}

type Input = { prompt: string; topics: string };

/**
 * Build the dataset + experiment (the part shared by the runnable `main` and the
 * parity test). `datasetName` is parameterized so tests can use a stable name;
 * `main` passes a timestamped one, exactly like the Java example.
 */
export function buildTopicRelevance(
  client: ExperimentsClient,
  datasetName = `topic-relevance-demo-${Date.now()}`,
): { dataset: Dataset; experiment: Experiment<Input, Record<string, unknown>> } {
  const dataset = client
    .createDataset(datasetName, "Java SDK v0.1 smoke test")
    .addRecord(
      { prompt: "I love hiking in the mountains on weekends.", topics: "outdoor, travel" },
      "true",
      { source: "synthetic", difficulty: "easy" },
    )
    .addRecord(
      { prompt: "Explain quantum entanglement in two sentences.", topics: "outdoor, travel" },
      "false",
      { source: "synthetic", difficulty: "easy" },
    )
    .addRecord(
      { prompt: "Best Italian restaurants in Brooklyn?", topics: "food, nyc" },
      "true",
      { source: "user-report", difficulty: "medium", reviewer: "alex" },
    )
    .addRecord(
      { prompt: "How do I configure nginx for HTTPS?", topics: "food, nyc" },
      "false",
      { source: "user-report", difficulty: "hard", reviewer: "alex" },
    );

  const experiment = Experiment.builder<Input, Record<string, unknown>>(client)
    .name("topic-relevance-demo")
    .dataset(dataset)
    .task((input) => {
      const overlap = keywordOverlap(input.prompt, input.topics);
      return { response: String(overlap), confidence: overlap ? 0.85 : 0.65 };
    })
    // boolean: did the predicted response match the expected label?
    .evaluator("exact_match", (_input, output, expected) => (output as any).response === expected)
    // score: the model's confidence
    .evaluator("confidence_score", (_input, output) => Number((output as any).confidence))
    // categorical: human-readable verdict
    .evaluator("verdict_category", (_input, output) =>
      (output as any).response === "true" ? "in-topic" : "off-topic",
    )
    .config({ approach: "keyword-overlap", version: "v0.1" })
    .tags({ variant: "java-v0.1", owner: "design-partner-bootstrap" })
    .build();

  return { dataset, experiment };
}

async function main(): Promise<void> {
  const client = new ExperimentsClient({
    apiKey: requireEnv("DD_API_KEY"),
    applicationKey: requireEnv("DD_APPLICATION_KEY"),
    site: process.env.DD_SITE ?? "datadoghq.com",
    projectName: "node-sdk-bootstrap",
  });

  const { dataset, experiment } = buildTopicRelevance(client);
  const result = await experiment.run();

  console.log(`Dataset URL    : ${dataset.url()}`);
  console.log(`Experiment URL : ${result.url}`);
  console.log(`Experiment ID  : ${result.experimentId}`);
  console.log(`Rows           : ${result.rows.length}`);
  for (const row of result.rows) {
    console.log(
      `  row ${row.index} status=${row.isError ? "error" : "ok"} evals=${JSON.stringify(row.evaluations)}`,
    );
  }
}

// Run only when invoked directly (not when imported by the parity test).
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
